##### [< Forrige](2%20-%20Python_Basics.ipynb)     |     [Neste >](4%20-%20lister_oppslag_numpy.ipynb)

<span style="color:#9bb1d4;font-size:30px;font-weight:900"> 3 - funksjoner </span>  

# Å bruke funksjoner

Funksjoner er helt sentralt i all slags programmering. Et program uten egendefinerte funksjoner er bare en liste med kommandoer, og det er svært begrenset hva du kan få til med det. Heldigvis er det ekstremt enkelt å lage egne funksjoner i Python. 

En vanlig nybegynnerfeil i programmering er å ikke utnytte muligheten til å lage egne funksjoner tilstrekkelig. Hovedregelen er at om det er noe av koden din som repeteres, så skal du lage en ny funksjon som håndterer dette. Målet er at antall linjer i hver funksjon skal være så lite som mulig. Men før vi lærer å lage egne funksjoner, skal vi se litt på hvordan noen innebygde funksjoner i Python brukes.

## Innebygde funksjoner

Python kommer med en rekke innebygde funksjoner. I tillegg kan du importere pakker som har en lang rekke funksjoner. Her er noen innebygde funksjoner i Python:

#### Eksempel 1:

In [16]:
# Finner maksimum av to tall
print(max(5, 3.0))

# Finner minimum av to tall
print(min(2.4, 2.0))

# Finner absoluttverdien av et tall
print(abs(-5))

# Finner lengden av en string.
print(len('jadda!'))

5
2.0
5
6


Legg merke til at funksjonene `max`, `min`, `abs` og `len` alle er omsluttet av funksjonen `print` som sørger for at resultatet skrives ut nedenfor. En funksjon kan altså *returnere noe*, slik som  `max`, `min`, `abs` og `len`, eller *gjøre noe* slik som `print`, eller begge deler.

Funksjonen `len` angir antall bokstaver, men kan også angi lengde på en liste. En liste i python skriver du som `[1,'hei',a]`. En liste kan altså bestå av flere ulike typer variabler og konstanter. De to første elementene her er konstanter, og den tredje er en variabel som er satt lik et flyttal i eksemplet under.

#### Eksempel 2:

In [17]:
alder = 27
print([1,'hei',27.4])
print(len([1,'hei',27.4]))


[1, 'hei', 27.4]
3


## Eksponentialfunksjonene

I Python er det svært enkelt å *importere* ferdig pakker for som kan kjøre ferdige funksjoner. Vi skal her se på et eksempel med noe som heter eksponentialfunksjonen. Denne finnes blant annet i pakken `numpy`. Dette er en svært nyttig pakke for matematiske beregninger. 

Som vi skal se lenger ned, så er eksponentialfunksjonen viktig innen økonomifaget fordi den brukes i vekstberegninger, for eksempel for befolkning eller produksjon. 

Som nevnt tidligere så er *potensen* $y^x$ et *grunntall* $y$ opphøyet i en eksponent $x$. Når vi snakker om *eksponenten* i bestemt form, så betyr det at grunntallet er et bestemt tall. Dette tallet kalles $e$, etter matematikeren Euler, og er tilnærmet $e\approx2.71828$ (det er egentlig en uendelig rekke med siffer etter komma) 

Siden Python ikke har noen innebygd funksjon for eksponenten, importerer vi altså først `numpy`-pakken, som har en slik funksjon. For å få mindre skriving senere, er det vanlig å importere `numpy` som aliaset `np`. Dette gjør vi slik:

#### Eksempel 3:

In [18]:
!pip install numpy

Defaulting to user installation because normal site-packages is not writeable


In [20]:
import numpy as np

In [21]:
print(np.__version__)

1.26.4


Navnet på eksponent-funksjonen i numpy er exp. Skal vi for eksempel finne hva $e^2=2.71828^2$ er, skriver vi

#### Eksempel 4:

In [22]:
np.exp(2)

7.38905609893065

 Om vi ønsker å vite hva eksponenten er med litt mer enn fem desimaler, kan vi finne eksponenten av 1, siden ethvert tall opphøyd i 1 er seg selv:

#### Eksempel 5:

In [23]:
np.exp(1)

2.718281828459045

Vi kan for eksempel skrive 2.71828 opphøyet i 5:

#### Eksempel 6:

In [24]:
np.exp(1)**5

148.41315910257657

In [25]:
np.exp(5)

148.4131591025766

For å opphøye noe bruker vi altså dobbelt gangetegn `**`.

*Logartimen* er det eksakt motsatte av eksponentialfunksjonen. Tar vi logaritmen av eksponentialfunksjon, får vi argumentet til sistnevnte. Tar vi eksponentialfunksjonen av logartimen får vi også argumentet til logartimen. I en av oppgavene under skal du vise dette.

Her har vi bare vist noen eksempler av ```exp```-funksjonen til biblioteket *Numpy*. Vi skal komme tilbake til Numpy senere i kurset, nå forsetter vi med å se på hvordan vi kan lage egne funksjoner.

# Lage egne funksjoner - renteregning

Vi har til nå sett på noen inneebygde-funksjoner i Python, nå skal vi se hvordan vi kan lage *egne* funksjoner. 

La oss ta for oss et eksempel med renteregning.

La oss starte med å lage en funksjon for et lån *uten* avdrag. Om banken legger til renten én gang i året og renten er ```r```, så er lånet på ```x``` kr om ```T``` år gitt som $x\cdot (1+r)^{T}$. Dette kan vi programmere som følgende funksjon:

#### Eksempel 7:

In [27]:
# Vi kan "hardkode" dette på følgende vis, med noen eksempelverdier

# Lån = 300000
# r = 0.1
# T = 3 år

300000*(1+0.1)**3

399300.0000000001

Her er det effektivt å lage en funksjon i stedet!

In [28]:
# Funksjonen 'account_balance' regner ut rente for lån uten avdrag
# x = lån
# r = rente
# T = år

def account_balance(x,r,T):
    return x*(1+r)**T

Her er ```account_balance``` *navnet* til funksjonen vår, og den tar inn tre *parametere* (nemlig ```x```, ```r``` og ```T```).

Nøkkelordet `return`sørger for at det som står etter `return` returneres av funksjonen. Dersom funksjonen ikke har `return`i seg, returneres `None`, som representerer "ingenting" i Python. 

Om renten på lånet er $20$% $= 0.2$ og innskuddet er $100$, så er beløpet om to år dermed:

#### Eksempel 8:

Når vi gir disse *parameterene* verdier, så kaller vi det for *argument*. 

Så parameterene her er: ```x```, ```r``` og ```T```

og argumentene i cellen under er: $100$, $0.2$ og $2$.

In [29]:
account_balance(100,0.2,2)

144.0

Det betyr altså at når innskuddet er $100$ og renta er på $20$% over $2$ år, øker lånet til $144$.

La oss nå tenke oss at banken legger til renter n ganger i året. Om for eksempel $n=12$, så legger banken til renter måndedlig. Vi kan da regne ut lånet ved ulik antall forrentninger per år. Formelen for dette er $x\cdot (1+\frac{r}{n})^{T\cdot n}$. 

Formelen har en enkel og intuitiv forklaring: Vi deler rentesatsen på antall perioder, samtidig som renten legges til i hver periode. Renten i hver periode er dermed $\frac{r}{n}$, og dette forrentes totalt $T\cdot n$ ganger. Dette kan programmeres i pyton som

#### Eksempel 9:

In [2]:
def account_balance_n(x,r,T,n):
    return x*(1+r/n)**(T*n)

La oss nå sammenligne å legge til renten hver måned, med å legge den til årlig:

#### Eksempel 10:

In [31]:
print(account_balance_n(100,0.2,2,12))
print(account_balance_n(100,0.2,2,1))

148.69146179463576
144.0


Vi ser at det lønner seg en god del å legge til renten hver måned. La oss nå anta at rentene legges til hvert sekund i året. Det er 31 536 000 sekunder i et år. I så fall er innskuddet om to år verdt:

#### Eksempel 11:

In [32]:
account_balance_n(100,0.2,2,31536000)

149.18246858790133

Banken vil altså tjene litt på å gjøre det. I stedet for å regne ut med daglig forrentning kan vi bruke kontinuerlig, som vil si at vi bruker eksponentialfunksjonen

#### Eksempel 12:

In [7]:
import numpy as np
def account_balance_exponential(x,r,T):
    return x*np.exp(r*T)

Dette gir oss verdien på lånet når renten legges til "kontinuerlig", altså oftere enn hvert sekund. La oss nå regne ut verdien av gjelden om to år med kontinuerlig forrentning:


#### Eksempel 13

In [8]:
account_balance_exponential(100,0.2,2)

149.18246976412703

Vi ser at dette er omtrent identisk med når vi la til renten hvert sekund. Eksponentialfunksjonen er altså en måte å regne ut vekst når økningen legges til veldig hyppig. Forrentning hvert sekund og kontinuerlig forrentig gir altså omtrent det samme resultatet. "Kontinuerlig forrentning" er et abstrakt begrep, men det betyr i realiteten bare at vi legger til veksten "veldig hyppig".  

Årsaken til at vi ofte bruker logaritmen og eksponentialfunksjonen istedet for å legge til veksten hvert sekund, ar et det er veldig mye enklere å regne med logartimer enn vanlig prosentregning. 

# Mer om funksjoner
La oss definere en enkel funksjon som dobler et tall:

#### Eksempel 14:

In [35]:
def f(x):
    return 2*x

 La oss nå sjekke følgende:

1. Hvilket resultat gir funksjonen?
2. Hva slags objekt er funksjonen?
3. Hva slags objekt returnerer funksjonen?

#### Eksempel 15:

In [36]:
print(f(8))            # printer resultatet av funksjonen når argumentet er 8
print(type(f))         # printer hvilken type objekt det er
print(type(f(5)))      # printer hvilken type resultate av funksjonen er

16
<class 'function'>
<class 'int'>


Funksjonen dobler argumentet, så med 8 som argument blir resultatet 16. Symbolet `f` representerer nå en funksjon (`function`), som vi ser er resultatet av `print(type(f))`. Men dersom vi *kaller* funksjonen med for eksemple argumentet 5, får vi resultatet av funksjonen, som er av type `int`. Funksjoner trenger imidlertid ikke retunere noe. Dette er også en funksjon:

#### Eksempel 16:

In [37]:
def hello():
    print('Hello, world!')

hello()

Hello, world!


Vi ser at denne funksjonen ikke har `return`-nøkkelordet. Dermed returneres ingen verdi. I stedet for å returnere noe, så *gjør* denne funksjonen noe, nemlig å skrive ut 'Hello world"' til skjermen. 

Du kaller funksjoner ved å skrive funksjonen med argumentet i parentes etter navnet. Som vi ser her, når det ikke er noen argumenter, skriver du tomme parenteser.

En funksjon er også en variabel og et objekt. Du kan for eksempel definere en ny variabel som en tidligere definert funksjon:

#### Eksempel 17:

In [38]:
hi=hello
hi()

Hello, world!


Av og til ønsker du at argumentene til en funksjon skal være valgfrie. Da kan du definere en standard ("default") verdi for argumentet:

#### Eksempel 18:

In [39]:
def f(x,a=1.02e+2):
    return a*x

print(f"Innehaveren krevde i utgangspunkltet {f(5)} kroner for fem epler.\n"
     f"Det var alt for mye, så jeg tilbydde {f(5,1.02e-2)} kroner.\n"
     f"Vi ble til slutt enige om {f(5,1.02e+1)} kroner.\n")

Innehaveren krevde i utgangspunkltet 510.0 kroner for fem epler.
Det var alt for mye, så jeg tilbydde 0.051000000000000004 kroner.
Vi ble til slutt enige om 51.0 kroner.



In [40]:
f(2)
f(2, 2)

4

Legg merke til at argumentene til en funksjon settes inn i en parentes. For `f(x,y)` for eksempel, så er `x` og `y` argumenter til funksjonen `f`. Disse står inne i en parentes `(x,y)`. Denne parentesen er et objekt i seg selv og kalles en `tuple`. Det går derfor an å definere alle argumentene *først*, og så la funksjonen evaluere dem. La oss for eksempel si at `x=20` og `y=50`, slik at `(x,y)=(20,50)`. Vi kan da definere variablene først og så evaluere funksjonen på følgende måte

In [41]:
args=(20,50)
f(*args)

1000

Dette er en mulighet som kan være veldig nyttig i mange sammenhenger. En `tuple`skiller seg fra en liste (`list`) ved at sistnevnte defineres med hakeparenteser `[]` i stedet for runde parenteser `()`. Vi skal senere komme tilbake til flere forskjeller på en liste og en tuple. 

# Tilbud og etterspørsel

La oss definere en tilbud- og etterspørselsfunksjon. Disse funksjonene skal vise hvilken ***pris*** konsumentene maksimalt er villige til å betale for en gitt mengde, og hvor mye produsentene minimum må ha for å selge varen. 

Den høyeste prisen er det bare noen få konsumenter som er villige til å betale, så desto større kvantum `x` som skal selges, desto lavere må prisen være. Etterspørselsfunksjonen må derfor være fallende (lavere pris ved høyere kvantum).

Her er et forslag til etterspørselsfunksjon:

#### Eksempel 19:

In [1]:
#Creating a function that returns total demand given a price p
def demand(x):
    return 125/(x+5)

Den laveste prisen er det bare den mest effektive bedriften som klarer å produsere lønnsomt for. Ettersom kvantum `x` øker vil flere, men mindre effektive bedrifter, tilby varen. Desto større kvantum, desto høyere pris kreves per enhet. Tilbudsfunksjonen bør derfor være stigende (høyere pris ved høyere kvantum).

Her er forslag til tilbudsfunksjoner:

#### Eksempel 20:

In [2]:
#Creating a function that returns total supply given a price p
def supply(x):
    return x**2

## Oppgaver:

#### Oppgave 1:

Sjekk med numeriske eksempler at tilbudskurven faktisk er økende og etterspørselskurven faktisk er fallende.

*Du lager en ny celle du kan kjøre kode i, ved å trykke "b" mens denne cellen er markert (enkeltklikk på denne cellen, ikke dobbelklikk)*

In [11]:
import numpy as np
x1 = np.arange(0,10)
print(x1)
print(demand(x1))

[0 1 2 3 4 5 6 7 8 9]
[25.         20.83333333 17.85714286 15.625      13.88888889 12.5
 11.36363636 10.41666667  9.61538462  8.92857143]


In [14]:
print(supply(10))
print(supply(x1))

100
[ 0  1  4  9 16 25 36 49 64 81]


#### Oppgave 2:
Finn med prøving og feiling eller på annen måte hvilket kvantum som gir likhet mellom tilbud og etterspørsel. Hvilken pris gir slik kvantum?

In [27]:
for i in np.arange(0,10,0.01):
    if (supply(i)-demand(i))**2 < 0.01:
        ox = i
print(ox)
    

3.7800000000000002


In [ ]:
#Alt metode som ikke er pensum
from scipy.optimize import fsolve

def equil(x):
    return demand(x) - supply(x)

eq_price = fsolve(equil, 10)
print(f'Likevektspris: {eq_price[0]}')

Likevektspris: 3.7743883312334976


#### Oppgave 3:

1. Lag en funksjon som viser hvor mye penger du har igjen når du har brukt x kr av et budsjett på b kr

2. Lag en funksjon som returnerer hvor mye parkeringen kostet når du har stått en tidsperiode angitt av argumentet `time_period`, og en pris per time angitt av argumentet `p`. Funksjonen skal forvente at time_period er på formatet "hh:mm:ss" (en streng). Du kan trekke ut timer, minutter og sekunder fra en tekstreng på det foreslåtte formatet, ved å bruke koden `hh,mm,ss=int(timestr[0:2]),int(timestr[3:5]),int(timestr[6:])`. Du må  konvertere tiden til timer på desimalform for å regne ut kostnad. Det kan du gjøre med denne formelen: `tt_dec=hh+mm/60+ss/(60*60)`. 

3. Lag en funksjon som tar en liste med adjektiv som argument. Kall argumentet for  `adj_list`. Funksjonen skal returnere et tilfeldig element fra listen. Du kan for eksempel teste funksjonen med denne listen `['dyr', 'billig', 'pareto optimal', 'høy', 'lav', 'ineffektiv']`. For å plukke et tilfeldig element fra en liste `adj_list` med `4`elementer, skriver du `adj_list[np.random.randint(4)]`.  


In [28]:
def budget(x, b):
    if x > b:
        print('Du har sprengt budsjettet')
    else:
        print(f'Du har {b-x} kr igjen av budsjettet på {b} kr, når du har brukt {x} kr.')

print(budget(1000, 10000))

Du har 9000 kr igjen av budsjettet på 10000 kr, når du har brukt 1000 kr.
None


In [61]:
time_period = '02:30:20'
hh,mm,ss=int(time_period[0:2]),int(time_period[3:5]),int(time_period[6:])
print(f'{hh}, {mm} og {ss}')

2, 30 og 20


In [64]:
def park_kostnad(time_period, p):
    hh,mm,ss=int(time_period[0:2]),int(time_period[3:5]),int(time_period[6:])
    tt_dec=hh+mm/60+ss/(60*60)
    print(f'Parkeringen din koster {round(tt_dec*p,2)} for tiden: ' + time_period)
    
park_kostnad('02:30:20', 20)    

Parkeringen din koster 50.11 for tiden: 02:30:20


In [44]:
#Oppgave 3
print(np.random.randint(4))

2


In [53]:
#Oppg 3
adj_list = ['dyr', 'billig', 'pareto optimal', 'høy', 'lav', 'ineffektiv']
print(adj_list[np.random.randint(len(adj_list))])

høy


#### Oppgave 4:
a) Bruk eksponentialfunksjonen og log-funksjonen i numpy til å vise at log-funksjonen er det motsatte av eksponentialfunksjoenn. 

b) Regn ut hvor mange prosent 205,210,220,230 og 240 er av 200. Regn også ut logaritmen av 205/200, 210/200 ... osv. og kommenter resultatet. 


In [55]:
import numpy as np
#Oppg 4 a)
y1 = np.exp(5)
print(y1)
y2 = np.log(y1)
print(y2)

148.4131591025766
5.0


In [57]:
# Oppg 4 b)
x2 = [205, 210, 220, 230, 240]
for i in x2:
    print(f'{i} er {i/200*100} % av {200}')

205 er 102.49999999999999 % av 200
210 er 105.0 % av 200
220 er 110.00000000000001 % av 200
230 er 114.99999999999999 % av 200
240 er 120.0 % av 200


In [59]:
for i in x2:
    print(f' log av {i/200} er {np.log(i/200)}')

 log av 1.025 er 0.024692612590371414
 log av 1.05 er 0.04879016416943204
 log av 1.1 er 0.09531017980432493
 log av 1.15 er 0.13976194237515863
 log av 1.2 er 0.1823215567939546


#### Oppgave 5:

Lag en funksjon som returnerer det høyeste av funksjonene `account_balance_exponential` og `account_balance`. Er dette en nyttig funksjon?

In [10]:
def highest_account_balance(x,r,T,n):
    return max((
        account_balance_n(x,r,T,n),
        account_balance_exponential(x,r,T)
    ))
    
print(highest_account_balance(100, 0.2, 4, 1))
account_balance_exponential(100, 0.2, 4)

222.55409284924679


222.55409284924679

# Oppgave 6

a) Bruk innebygde Python-funksjoner til å finne maksimum, minimum, absoluttverdi og lengden av teksten "1, -5, 10". Bruk funksjonen `eval` til å konvertere verdiene til en tallrekke. 

In [70]:
s1 = '1,-5,10'
print(f'Lengden av s1 er {len(s1)}')
x3 = eval(s1)
print(x3)
print(max(x3))
print(min(x3))
print(np.abs(x3))

Lengden av s1 er 7
(1, -5, 10)
10
-5
[ 1  5 10]


### Nåverdi

b) Nåverdien til et beløp er det en trenger å sette i banken i dag for å få beløpet i fremtiden. Formelen for dette er `X/((1+r)**T)`. Lag en funksjon som retuner nåverdien av et beløp. Lag også en funksjon som returnerer nåverdien med kontinuerlig forrentning.

c) Lag en funkson som returnerer nåverdi eller sluttverdi avhengig av hva argumentene er. Sluttverdi er det som er beregnet i avsnittet "Lage egne funksjoner - renteregning". Funksjonen skal ha antall forrentninger per år som argument, og skal returnere kontinuerlig forrentning dersom antall forrentninger er høyere enn 100

Du kan bruke en `if`- betingelse til å velge hva en funksjon skal gjøre slik:

In [71]:
#6b)
def naaverdi(T, r, X):
    print(f'Nåverdien til {X} ved rente {r} er: {X/((1+r)**T)}')
    print(f'Nåverdien til {X} ved rente {r} er: {X*np.exp(-r*T)}')

naaverdi(10,0.05,10000)

Nåverdien til 10000 ved rente 0.05 er: 6139.132535407592
Nåverdien til 10000 ved rente 0.05 er: 6065.306597126334


In [73]:
#6c)

def naaverdi_sluttverdi(T, r, X, navd = True):
    if navd and T <= 100:
        print(f'Nåverdien til {X} ved rente {r} er: {X/((1+r)**T)}')
    elif navd and T > 100:
        print(f'Nåverdien til {X} ved rente {r} er: {X*np.exp(-r*T)}')
    elif navd == False  and T <= 100:
        print(f'Sluttverdien til {X} ved rente {r} er: {X*((1+r)**T)}')
    elif navd == False and T > 100:
        print(f'Sluttverdien til {X} ved rente {r} er: {X*np.exp(r*T)}')

navd = False
naaverdi_sluttverdi(10,0.05,10000,navd)
    
        
        


Sluttverdien til 10000 ved rente 0.05 er: 16288.94626777442


### Men vi kan også bruke at med negativ tid T får vi nåverdi, slik at vi slipper et eget argument for det

##### [< Forrige](2%20-%20Python_Basics.ipynb)     |     [Neste >](4%20-%20lister_oppslag_numpy.ipynb)